src/hardware/opl.cpp

```diff
@@ -511,11 +512,50 @@ io_port_t OPL::WriteAddr(const io_port_t port, const uint8_t val)
 	return addr;
 }
 
+template <LineIndex line_index>
+int16_t remove_dc_bias(const int16_t in_sample)
+{
+	constexpr int16_t BiasThreshold = 5;
+	constexpr size_t BiasBlockSize  = 16;
+
+	static std::deque<int16_t> samples = {};
+
+	// Clear the queue if the stream falls back to zero
+	if (in_sample < BiasThreshold) {
+		samples.clear();
+		return in_sample;
+	}
+
+	// Otherwise start adding to the queue
+	samples.push_back(in_sample);
+	if (samples.size() < BiasBlockSize) {
+		return {};
+	}
+
+	// Compute the running bias
+	int32_t sum = 0;
+	for (const auto s : samples) {
+		sum += s;
+	}
```

Comment:
Probably wouldn't make a huge difference, but you could use a moving average approach instead. So you wouldn't need to average 2 * 16 samples on each invocation. Example implementtaion:

https://www.zrzahid.com/moving-average-of-last-n-numbers-in-a-stream/

Reply:
Thanks for the tip; will swap this in tomorrow 👍

Reply:
I think you could also get rid of the conditionals completely and make it branchless. More for fun & simplicity than the actual performance benefits 😄 Just maintain a running average of the DC offset that starts from zero, taking the last N=16 samples into account.

Reply:
Replaced it with a simple running average:

```c++
int16_t remove_running_average(const int16_t sample)
{
	// Calculate the number of samples we need average across to maintain
	// the lowest frequency given an assumed playback rate.
	constexpr int16_t PcmPlaybackRateHz      = 16000;
	constexpr int16_t LowestFreqToMaintainHz = 200;
	constexpr int16_t NumToAverage = PcmPlaybackRateHz / LowestFreqToMaintainHz;

	static int32_t sum = 0;
	static std::queue<int16_t> samples = {};

	sum += sample;
	samples.push(sample);

	int16_t average = 0;
	if (samples.size() == NumToAverage) {
		average = static_cast<int16_t>(sum / NumToAverage);
		sum -= samples.front();
		samples.pop();
	}
	return static_cast<int16_t>(sample - average);
```

However, a bunch of the pops remain:

https://github.com/dosbox-staging/dosbox-staging/assets/1557255/83e95e2e-91f1-4ed0-b8d7-4623558df640

I've added this running-average approach as the first commit (just for the record), and then modified it in the second commit with the original algorithm.

Reply:
(resolving; the running average approach is now in the original algorithm)